storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,278 @@
+	"github.com/prometheus/tsdb/wal"
+	fsnotify "gopkg.in/fsnotify/fsnotify.v1"
+)
+
+// WALWatcher watches the TSDB WAL for a given StorageClient (remote write config).
+type WALWatcher struct {
+	storage *Storage
+	logger  log.Logger
+
+	currentSegment  *wal.Segment
+	walDir          string
+	lastForwardedTs prometheus.Gauge
+
+	dirChan        chan string
+	ctx            context.Context
+	cancel         context.CancelFunc
+	quit           chan struct{}
+	exitForwarding chan struct{}
+}
+
+// NewWALWatcher creates a new WAL watcher for a given remote write client.
+func NewWALWatcher(logger log.Logger, storage *Storage, dirChan chan string) *WALWatcher {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	w := &WALWatcher{
+		logger:         logger,
+		storage:        storage,
+		dirChan:        dirChan,
+		ctx:            ctx,
+		cancel:         cancel,
+		quit:           make(chan struct{}),
+		exitForwarding: make(chan struct{}),
+	}
+	w.lastForwardedTs = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "prometheus_remote_write_last_forwarded_timestamp",
+		Help: "Timestamp of the last time the WAL watcher sent a sample to queue managers.",
+	})
+	prometheus.MustRegister(w.lastForwardedTs)
+	return w
+}
+
+func (w *WALWatcher) setWALDir(dir string) {
+	w.logger.Log("msg", fmt.Sprintf("setting wal dir to %s", dir))
+	w.walDir = dir
+}
+
+func (w *WALWatcher) Start() {
+	go w.watch()
+}
+
+func (w *WALWatcher) Stop() {
+	// It seems that sometimes the quit channel is already closed.
+	defer func() {
+		if r := recover(); r != nil {
+			level.Error(log.With(w.logger)).Log("err", r)
+		}
+	}()
+	close(w.exitForwarding)
+	close(w.quit)
+	return
+}
+
+// When starting the WAL watcher, there is potentially an existing WAL. In that case, we
+// should read to the end of the latest segment before listening for file events.
+func (w *WALWatcher) readToEnd(s *wal.Segment) {
+	r := wal.NewReader(s)
+	for r.Next() {
+	}
+}
+
+func (w *WALWatcher) watch() {
+	// The WAL dir won't be set when Prometheus first starts up, wait to recv.
+	w.walDir = <-w.dirChan
+
+	var err error
+	nw, err := wal.New(nil, nil, w.walDir)
+	if err != nil {
+		w.logger.Log("err", err)
+	}
+	_, n, err := nw.Segments()
+
+	if err != nil {
+		w.logger.Log("err", err)
+	}
+	if n == -1 {
+		w.logger.Log("err", "no segments found")
+		return
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		w.logger.Log("err", err)
+		return
+	}
+
+	defer watcher.Close()
+	// Open the current segment.
+	var segment *wal.Segment
+	segment, err = wal.OpenReadSegment(wal.SegmentName(w.walDir, n))
+	if err != nil {
+		w.logger.Log("err", err)
+		return
+	}
+	w.readToEnd(segment)
+	// Send a message back to signal that we've read to the end of the current WAL segment and are starting to watch
+	// writes and new WAL segment files.
+	w.dirChan <- ""
+	var currentSegment = fmt.Sprintf("%08d", n)
+	defer segment.Close()
```

Comment:
This `defer` will run when you exit, and captures the current value of `segment` I think.

Reply:
Put segment handling in a separate function to make closing semantics clear; then call that function in a loop.